libexec/revng/revng-check-conventions

```diff
@@ -1,498 +1,727 @@
+
+        return PassResult(
+            generate_error_lines("Whitespace at the end of line:", trailing_whitespace)
+            + generate_error_lines("Tabs present:", tabs)
+            + generate_error_files("Files that don't end in a newline:", no_end_newline),
+            empty(trailing_whitespace, tabs, no_end_newline),
+        )
+
+
+class RevngCPass(Pass):
+    def __init__(self):
+        super().__init__("revng-cpp-checks", "c")
+        self.generic_matchers = {
+            "There are lines longer than 80 characters:": regexes_matcher(r"^.{81,}$"),
+            "Found snippets that should not be present:": regexes_matcher(
+                r"\(--> 0\)", r";;", r"^\s*->.*;$", r"Twine [^&]"
+            ),
+            "Found snippets that shouldn't be at the end of a line:": regexes_matcher(
+                r"::$", r"<$", r"RegisterPass.*>$", r"} else$", r"\bopt\b.*>$"
+            ),
+            "Includes should never use <..> except for C++ standard includes:": regexes_matcher(
+                r"^\s*#include <.*\.hpp>"
+            ),
+            "Parenthesis at the end of line:": self.parethesis_matcher,
+            "Found snippets that should never be at the beginning of a line:": (
+                regexes_matcher_prefix(
+                    r"^\s*",
+                    r"\*>",
+                    r"/[^/\*]",
+                    r":[^:\(]*\)",
+                    r"==",
+                    r"\!=",
+                    r"<[^<]",
+                    r">",
+                    r">=",
+                    r"<=",
+                    r"//\s*WIP",
+                    r"#if\s*[01]",
+                )
+            ),
+        }
+
+        self.non_support_matchers = {
+            "Use revng_{assert,check,abort,unreachable}:": regexes_matcher(
+                r"\babort\(", r"\bassert\(", r"assert\(false", r"llvm_unreachable"
+            )
+        }
+
+        self.cpp_matchers = {}
+        self.header_matchers = {"Headers should not use static": regexes_matcher(r"^static\s")}
+
+    @staticmethod
+    def parethesis_matcher(_: Path, line: str) -> List[re.Match]:
+        if re.search(r"R\"LLVM.*\($", line):
+            return []
+        main_match = re.search(r"\($", line)
+        return [main_match] if main_match else []
+
+    @staticmethod
+    def include_matcher(cpp_files: Dict[Path, str]):
+        def include_matcher_inner(file: Path, line: str) -> List[re.Match]:
+            if file in cpp_files:
+                match = re.search(r"^\s*#include <.*\.h(|pp)>", line)
+            else:
+                match = re.search(r"^\s*#include <.*\.hpp>", line)
+            return [match] if match else []
+
+        return include_matcher_inner
+
+    def process(self, files: Dict[Path, str], args) -> PassResult:
+        non_support_files = {
+            path: content
+            for path, content in files.items()
+            if str(path) not in ["share/revng/support.c", "lib/Support/Assert.cpp"]
```

Comment:
rcc is sprinkled with hardcoded strings. It'd be nice to have a configuration file in YAML for as many things as possible.
If we don't feel like going this route right now, at least please concentrate all the configuration (in particular, hardcoded strings), in a single place.

When I'll get unhappy about rcc, it'd be nice to have a single place where to change the configuration (maybe with docs), instead of having to search what I want among the program logic.

Reply:
Great!